Mathlib/Algebra/Category/ModuleCat/Differentials/Presheaf.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf
+import Mathlib.Algebra.Category.ModuleCat.Differentials.Basic
+
+/-!
+# The presheaf of differentials of a presheaf of modules
+
+In this file, we define the type `M.Derivation φ` of derivations
+with values in a presheaf of `R`-modules `M` relative to
+a morphism of `φ : S ⟶ F.op ⋙ R` of presheaves of commutative rings
+over categories `C` and `D` that are related by a functor `F : C ⥤ D`.
+We formalize the notion of universal derivation.
+
+Geometrically, if `f : X ⟶ S` is a morphisms of schemes (or more generally
+a morphism of commutative ringed spaces), we would like to apply
+these definitions in the case where `F` is the pullback functors from
+open subsets of `S` to open subsets of `X` and `φ` is the
+morphism $O_S ⟶ f_* O_X$.
+
+In order to prove that there exists a universal derivation, the target
+of which shall be called the presheaf of relative differentials of `φ`,
+we first study the case where `F` is the identity functor.
+In this case where we have a morphism of presheaves of commutative
+rings `φ' : S' ⟶ R`, we construct a derivation
+`DifferentialsConstruction.derivation'` which is universal.
+Then, the general case (TODO) shall be obtained by observing that
+derivations for `S ⟶ F.op ⋙ R` identify to derivations
+for `S' ⟶ R` where `S'` is the pullback by `F` of the presheaf of
+commutative rings `S` (the data is the same: it suffices
+to show that the two vanishing conditions `d_app` are equivalent).
+
+-/
+
+universe v u v₁ v₂ u₁ u₂
+
+open CategoryTheory
+
+variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₂} D]
+
+namespace PresheafOfModules
+
+variable {S : Cᵒᵖ ⥤ CommRingCat.{u}} {F : C ⥤ D} {S' R : Dᵒᵖ ⥤ CommRingCat.{u}}
+   (M N : PresheafOfModules.{v} (R ⋙ forget₂ _ _))
+   (φ : S ⟶ F.op ⋙ R) (φ' : S' ⟶ R)
+
+/-- Given a morphism of presheaves of commutative rings `φ : S ⟶ F.op ⋙ R`,
+this is the type of relative `φ`-derivation of a presheaf of `R`-modules `M`. -/
+@[ext]
+structure Derivation where
+  /-- the underlying additive map `R.obj X →+ M.obj X` of a derivation -/
+  d {X : Dᵒᵖ} : R.obj X →+ M.obj X
+  d_mul {X : Dᵒᵖ} (a b : R.obj X) : d (a * b) = a • d b + b • d a := by aesop_cat
+  d_map {X Y : Dᵒᵖ} (f : X ⟶ Y) (x : R.obj X) :
+    d (R.map f x) = M.map f (d x) := by aesop_cat
+  d_app {X : Cᵒᵖ} (a : S.obj X) : d (φ.app X a) = 0 := by aesop_cat
```

Comment:
Can't we just use the newly introduced bundled derivation (in terms of `Module.Cat`) here?

Reply:
With this degree of generality, we cannot use bundled derivations, but in the specific case of `Derivation'`, I will try to add a constructor which takes a family of bundles derivations, and a constructor for `IsUniversal` also.